File: src/spark-project/sparder/src/main/java/io/kyligence/kap/query/pushdown/SparkSqlClient.java
/*
 * Copyright (C) 2016 Kyligence Inc. All rights reserved.
 *
 * http://kyligence.io
 *
 * This software is the confidential and proprietary information of
 * Kyligence Inc. ("Confidential Information"). You shall not disclose
 * such Confidential Information and shall use it only in accordance
 * with the terms of the license agreement you entered into with
 * Kyligence Inc.
 *
 * THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS
 * "AS IS" AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT
 * LIMITED TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR
 * A PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT
 * OWNER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL,
 * SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT
 * LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE,
 * DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY
 * THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
 * (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
 * OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
 */


/*
 * Licensed to the Apache Software Foundation (ASF) under one
 * or more contributor license agreements.  See the NOTICE file
 * distributed with this work for additional information
 * regarding copyright ownership.  The ASF licenses this file
 * to you under the Apache License, Version 2.0 (the
 * "License"); you may not use this file except in compliance
 * with the License.  You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package io.kyligence.kap.query.pushdown;

import com.google.common.base.Strings;
import com.google.common.collect.Lists;
import org.apache.kylin.common.KylinConfig;
import org.apache.kylin.common.exceptions.KylinTimeoutException;
import org.apache.kylin.common.util.Pair;
import org.apache.kylin.shaded.htrace.org.apache.htrace.Trace;
import org.apache.spark.api.java.JavaRDD;
import org.apache.spark.sql.Dataset;
import org.apache.spark.sql.Row;
import org.apache.spark.sql.SparkSession;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;
import scala.collection.JavaConversions;

import java.io.Serializable;
import java.sql.Types;
import java.util.ArrayList;
import java.util.HashMap;
import java.util.List;
import java.util.Map;
import java.util.UUID;
import java.util.concurrent.Semaphore;
import java.util.regex.Matcher;
import java.util.regex.Pattern;

public class SparkSqlClient implements Serializable {
    public static final Logger logger = LoggerFactory.getLogger(SparkSqlClient.class);
    public static final long serialVersionUID = 12L;
    private final transient SparkSession ss;
    private final transient Semaphore semaphore;
    private Map<UUID, Integer> uuidSizeMap = new HashMap<>();
    private final int maxDriverMem;
    private final int allocationTimeOut;

    public SparkSqlClient(SparkSession sc, Semaphore semaphore) {
        this.ss = sc;
        ss.sparkContext().setLocalProperty("spark.scheduler.pool", "query_pushdown");
        this.semaphore = semaphore;
        this.maxDriverMem = semaphore.availablePermits();
        this.allocationTimeOut = Strings
                .isNullOrEmpty(System.getProperty("kap.storage.columnar.driver-allocation-timeout")) ? 60
                        : Integer.parseInt(System.getProperty("kap.storage.columnar.driver-allocation-timeout"));
    }

    public Pair<List<List<String>>, List<StructField>> executeSql(String sql, UUID uuid) {
        String s = "Start to run sql with SparkSQL...";
        logger.info(s);
        Trace.addTimelineAnnotation(s);

        //Get result data
        Dataset<Row> df = ss.sql(sql);

        String msg = "SparkSQL returned result DataFrame";
        logger.info(msg);

        Trace.addTimelineAnnotation(msg);
        return DFToList(uuid, df);
    }

    private Pair<List<List<String>>, List<StructField>> DFToList(UUID uuid, Dataset<Row> df) {
        String jobGroup = Thread.currentThread().getName();
        ss.sparkContext().setJobGroup(jobGroup, "Push down: " , true);
        try {
            JavaRDD<List<String>> rowRdd = df.javaRDD().mapPartitions(new FlatMapFunctionImpl());

            //Get struct fields
            List<List<String>> sampleList;

            sampleList = rowRdd.takeSample(true, 1);

            long sampleLen = 0;

            if (!sampleList.isEmpty()) {
                for (String elem : sampleList.get(0)) {
                    if (elem != null) {
                        sampleLen += elem.getBytes().length;
                    }
                }
            }

            int estimateSize = (int) Math.ceil((double) (rowRdd.count() * sampleLen) / (1024 * 1024));
            uuidSizeMap.put(uuid, estimateSize);

            if (estimateSize > this.maxDriverMem) {
                throw new RuntimeException("Estimate size exceeds the maximum driver memory size");
            }

            /*try {
                if (!this.semaphore.tryAcquire(estimateSize, allocationTimeOut, TimeUnit.SECONDS)) {
                    String format = String.format(
                            "spark driver allocation memory more than %s s," + "please increase driver memory or"
                                    + " set kap.storage.columnar.driver-allocation-timeout ",
                            allocationTimeOut);
                    throw new RuntimeException(format);
                }
            } catch (InterruptedException e) {
                throw new RuntimeException("interruptted", e);
            }*/

            logger.info("Estimate size of dataframe is " + estimateSize + "m.");

            List<org.apache.spark.sql.types.StructField> originFieldList = JavaConversions.seqAsJavaList(df.schema());
            List<StructField> fieldList = new ArrayList<>(originFieldList.size());

            for (org.apache.spark.sql.types.StructField field : originFieldList) {
                StructField.StructFieldBuilder builder = new StructField.StructFieldBuilder();
                builder.setName(field.name());

                int type = Types.OTHER;
                String typeName = field.dataType().sql();

                switch (typeName) {
                case "BINARY":
                    type = Types.BINARY;
                    break;
                case "BOOLEAN":
                    type = Types.BOOLEAN;
                    break;
                case "DATE":
                    type = Types.DATE;
                    break;
                case "DOUBLE":
                    type = Types.DOUBLE;
                    break;
                case "FLOAT":
                    type = Types.FLOAT;
                    break;
                case "INT":
                    type = Types.INTEGER;
                    break;
                case "BIGINT":
                    type = Types.BIGINT;
                    break;
                case "NUMERIC":
                    type = Types.NUMERIC;
                    break;
                case "SMALLINT":
                    type = Types.SMALLINT;
                    break;
                case "TIMESTAMP":
                    type = Types.TIMESTAMP;
                    break;
                case "STRING":
                    type = Types.VARCHAR;
                    break;
                default:
                    break;
                }

                if (typeName.startsWith("DECIMAL")) {
                    Pair<Integer, Integer> precisionAndScalePair = getDecimalPrecisionAndScale(typeName);

                    if (precisionAndScalePair != null) {
                        builder.setPrecision(precisionAndScalePair.getFirst());
                        builder.setScale(precisionAndScalePair.getSecond());
                    }
                    type = Types.DECIMAL;
                    typeName = "DECIMAL";
                }
                builder.setDataType(type);
                builder.setDataTypeName(typeName);
                builder.setNullable(field.nullable());

                fieldList.add(builder.createStructField());
            }
            List<List<String>> rowList = Lists.newArrayList();
            /*if (SparderEnv.isAsyncQuery()) {
                SparderEnv.getResultRef().set(true);
                SparderEnv.setDF(df);
                final String separator = SparderEnv.getSeparator();
                String path = KapConfig.getInstanceFromEnv().getAsyncResultBaseDir(QueryContext.current().getProject()) + "/"
                        + QueryContext.current().getQueryId();
                JavaRDD<String> mapRdd = rowRdd.map(new Function<List<String>, String>() {
                    @Override
                    public String call(List<String> v1) {
                        return StringUtil.join(v1, separator);
                    }
                });
                if (KapConfig.getInstanceFromEnv().isAsyncResultRepartitionEnabled()) {
                    mapRdd.repartition(SparderEnv.getAsyncResultCore()).saveAsTextFile(path);
                } else {
                    mapRdd.saveAsTextFile(path);
                }
            } else {
                rowList = rowRdd.collect();
            }*/
            rowList = rowRdd.collect();
            rowRdd.unpersist();
            df.unpersist();
            return Pair.newPair(rowList, fieldList);

        } catch (Throwable e) {
            if (e instanceof InterruptedException) {
                ss.sparkContext().cancelJobGroup(jobGroup);
                logger.info("Query timeout ", e);
                Thread.currentThread().interrupt();
                throw new KylinTimeoutException(
                        "Query timeout after: " + KylinConfig.getInstanceFromEnv().getQueryTimeoutSeconds() + "s");
            } else {
                throw e;
            }
        }
    }

    private Pair<Integer, Integer> getDecimalPrecisionAndScale(String type) {
        Pattern DECIMAL_PATTERN = Pattern.compile("DECIMAL\\(([0-9]+),([0-9]+)\\)", Pattern.CASE_INSENSITIVE);
        Matcher decimalMatcher = DECIMAL_PATTERN.matcher(type);
        if (decimalMatcher.find()) {
            int precision = Integer.valueOf(decimalMatcher.group(1));
            int scale = Integer.valueOf(decimalMatcher.group(2));

            return new Pair<>(precision, scale);
        } else {
            return null;
        }
    }

}

File: src/spark-project/sparder/src/main/scala/io/kyligence/kap/query/pushdown/SparkSqlClient.scala
/*
 * Copyright (C) 2016 Kyligence Inc. All rights reserved.
 *
 * http://kyligence.io
 *
 * This software is the confidential and proprietary information of
 * Kyligence Inc. ("Confidential Information"). You shall not disclose
 * such Confidential Information and shall use it only in accordance
 * with the terms of the license agreement you entered into with
 * Kyligence Inc.
 *
 * THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS
 * "AS IS" AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT
 * LIMITED TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR
 * A PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT
 * OWNER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL,
 * SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT
 * LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE,
 * DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY
 * THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
 * (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
 * OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
 */

package io.kyligence.kap.query.pushdown

import io.kyligence.kap.guava20.shaded.common.collect.Lists
import io.kyligence.kap.metadata.project.NProjectManager
import io.kyligence.kap.metadata.query.StructField
import io.kyligence.kap.query.mask.QueryResultMasks
import io.kyligence.kap.query.runtime.plan.QueryToExecutionIDCache
import io.kyligence.kap.query.runtime.plan.ResultPlan.saveAsyncQueryResult
import io.kyligence.kap.query.util.SparkJobTrace
import org.apache.commons.lang3.StringUtils
import org.apache.kylin.common.exception.KylinTimeoutException
import org.apache.kylin.common.util.{DateFormat, HadoopUtil, Pair}
import org.apache.kylin.common.{KapConfig, KylinConfig, QueryContext}
import org.apache.kylin.query.SlowQueryDetector
import org.apache.kylin.query.exception.UserStopQueryException
import org.apache.spark.network.util.JavaUtils
import org.apache.spark.sql.hive.QueryMetricUtils
import org.apache.spark.sql.hive.utils.ResourceDetectUtils
import org.apache.spark.sql.util.SparderTypeUtil
import org.apache.spark.sql.{DataFrame, SparderEnv, SparkSession}
import org.slf4j.{Logger, LoggerFactory}
import java.sql.Timestamp
import java.util.{UUID, List => JList}

import com.google.common.collect.ImmutableList

import scala.collection.JavaConverters._
import scala.collection.{immutable, mutable}

object SparkSqlClient {
  val DEFAULT_DB: String = "spark.sql.default.database"

  val logger: Logger = LoggerFactory.getLogger(classOf[SparkSqlClient])

  @deprecated
  def executeSql(ss: SparkSession, sql: String, uuid: UUID, project: String): Pair[JList[JList[String]], JList[StructField]] = {
    val results = executeSqlToIterable(ss, sql, uuid, project)
    Pair.newPair(ImmutableList.copyOf(results._1), results._3)
  }

  def executeSqlToIterable(ss: SparkSession, sql: String, uuid: UUID, project: String):
  (java.lang.Iterable[JList[String]], Int, JList[StructField]) = {
    ss.sparkContext.setLocalProperty("spark.scheduler.pool", "query_pushdown")
    HadoopUtil.setCurrentConfiguration(ss.sparkContext.hadoopConfiguration)
    val s = "Start to run sql with SparkSQL..."
    val queryId = QueryContext.current().getQueryId
    ss.sparkContext.setLocalProperty(QueryToExecutionIDCache.KYLIN_QUERY_ID_KEY, queryId)
    logger.info(s)

    try {
      val db = if (StringUtils.isNotBlank(project)) {
        NProjectManager.getInstance(KylinConfig.getInstanceFromEnv).getDefaultDatabase(project)
      } else {
        null
      }
      ss.sessionState.conf.setLocalProperty(DEFAULT_DB, db)
      val df = QueryResultMasks.maskResult(ss.sql(sql))

      autoSetShufflePartitions(df)

      val msg = "SparkSQL returned result DataFrame"
      logger.info(msg)

      dfToList(ss, sql, df)
    } finally {
      ss.sessionState.conf.setLocalProperty(DEFAULT_DB, null)
    }
  }

  private def autoSetShufflePartitions(df: DataFrame) = {
    val config = KylinConfig.getInstanceFromEnv
    if (config.isAutoSetPushDownPartitions) {
      try {
        val basePartitionSize = config.getBaseShufflePartitionSize
        val paths = ResourceDetectUtils.getPaths(df.queryExecution.sparkPlan)
        val sourceTableSize = ResourceDetectUtils.getResourceSize(config.isConcurrencyFetchDataSourceSize,
          paths: _*) + "b"
        val partitions = Math.max(1, JavaUtils.byteStringAsMb(sourceTableSize) / basePartitionSize).toString
        df.sparkSession.sessionState.conf.setLocalProperty("spark.sql.shuffle.partitions", partitions)
        QueryContext.current().setShufflePartitions(partitions.toInt)
        logger.info(s"Auto set spark.sql.shuffle.partitions $partitions, " +
          s"sourceTableSize $sourceTableSize, basePartitionSize $basePartitionSize")
      } catch {
        case e: Throwable =>
          logger.error("Auto set spark.sql.shuffle.partitions failed.", e)
      }
    }
  }

  /* VisibleForTesting */
  def dfToList(ss: SparkSession, sql: String, df: DataFrame): (java.lang.Iterable[JList[String]], Int, JList[StructField]) = {
    val config = KapConfig.getInstanceFromEnv
    val jobGroup = Thread.currentThread.getName
    ss.sparkContext.setJobGroup(jobGroup, s"Push down: $sql", interruptOnCancel = true)
    try {
      val queryTagInfo = QueryContext.current().getQueryTagInfo
      if (queryTagInfo.isAsyncQuery) {
        saveAsyncQueryResult(df, queryTagInfo.getFileFormat, queryTagInfo.getFileEncode, null)
        val fieldList = df.schema.map(field => SparderTypeUtil.convertSparkFieldToJavaField(field)).asJava
        return (Lists.newArrayList(), 0, fieldList)
      }
      QueryContext.currentTrace().endLastSpan()
      val jobTrace = new SparkJobTrace(jobGroup, QueryContext.currentTrace()
        , QueryContext.current().getQueryId, SparderEnv.getSparkSession.sparkContext)
      val results = df.toIterator()
      val resultRows = results._1
      val resultSize = results._2

      if (config.isQuerySparkJobTraceEnabled) jobTrace.jobFinished()
      val fieldList = df.schema.map(field => SparderTypeUtil.convertSparkFieldToJavaField(field)).asJava
      val (scanRows, scanBytes) = QueryMetricUtils.collectScanMetrics(df.queryExecution.executedPlan)
      QueryContext.current().getMetrics.setScanRows(scanRows)
      QueryContext.current().getMetrics.setScanBytes(scanBytes)

      (
        () => new java.util.Iterator[JList[String]] {
          override def hasNext: Boolean = resultRows.hasNext

          override def next(): JList[String] = {
            resultRows.next().toSeq.map(rawValueToString(_)).asJava
          }
        },
        resultSize,
        fieldList
      )
    } catch {
      case e: Throwable =>
        if (e.isInstanceOf[InterruptedException]) {
          Thread.currentThread.interrupt()
          ss.sparkContext.cancelJobGroup(jobGroup)
          if (SlowQueryDetector.getRunningQueries.get(Thread.currentThread()).isStopByUser) {
            throw new UserStopQueryException("")
          }
          QueryContext.current.getQueryTagInfo.setTimeout(true)
          logger.info("Query timeout ", e)
          throw new KylinTimeoutException("The query exceeds the set time limit of "
            + KylinConfig.getInstanceFromEnv.getQueryTimeoutSeconds + "s. Current step: Collecting dataset for push-down. ")
        }
        else throw e
    } finally {
      QueryContext.current().setExecutionID(QueryToExecutionIDCache.getQueryExecutionID(QueryContext.current().getQueryId))
      df.sparkSession.sessionState.conf.setLocalProperty("spark.sql.shuffle.partitions", null)
      HadoopUtil.setCurrentConfiguration(null)
    }
  }

  private def rawValueToString(value: Any, wrapped: Boolean = false): String = value match {
    case null => null
    case value: Timestamp => DateFormat.castTimestampToString(value.getTime)
    case value: String => if (wrapped) "\"" + value + "\"" else value
    case value: mutable.WrappedArray.ofRef[Any] => value.array.map(v => rawValueToString(v, true)).mkString("[", ",", "]")
    case value: immutable.Map[Any, Any] =>
      value.map(p => rawValueToString(p._1, true) + ":" + rawValueToString(p._2, true)).mkString("{", ",", "}")
    case value: Any => value.toString
  }
}

class SparkSqlClient
